iceoryx_hoofs/README.md

```diff
@@ -39,112 +22,129 @@ and we do not use dynamic memory. In these cases we adjusted the API to our use
+|`smart_lock`           | i        | Creates arbitrary thread-safe constructs which then can be used like smart pointers. If some STL type should be thread safe use the smart_lock to create the thread safe version in one line. Based on some ideas presented in [Wrapping C++ Member Function Calls](https://stroustrup.com/wrapper.pdf) |
+|`mutex`                | i        | Mutex interface, see [ManPage pthread_mutex_lock](https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html).                                                                                                                  |
+|`UnnamedSemaphore`     |          | Unamed semaphore interface, see [ManPage sem_overview](https://man7.org/linux/man-pages/man7/sem_overview.7.html)                                                                                                                     |
+|`NamedSemaphore`       |          | Named semaphore interface, see [ManPage sem_overview](https://man7.org/linux/man-pages/man7/sem_overview.7.html)                                                                                                                      |
+|`thread`               |          | Heap-less replacement for `std::thread`.                                                                                                                                                                                              |
+|`SignalGuard`          |          | Helper class for signal handler registration.                                                                                                                                                                                         |
+
+### Generalized design patterns & abstractions (design)
+
+| class                 | internal | description                                                         |
+|:---------------------:|:--------:|:--------------------------------------------------------------------|
+|`Builder`              |          | Macro which generates a setter method useful for a builder pattern. |
+|`posixCall`            |          | Wrapper around C and POSIX function calls which performs a full error handling. Additionally, this wrapper makes sure that `EINTR` handling is performed correctly by repeating the system call. |
+|`functional_interface` |          | Constructs to easily add functional interfaces like `and_then` to object container.                                                                                                                                                   |
+|`NewType<T, Policies>` |          | C++11 implementation of [Haskells NewType-pattern](https://wiki.haskell.org/Newtype).                                                                                                                                                 |
 
-### Error handling
+### Error handling & logging (error)
```

Comment:
How about calling this `Reporting (reporting)`. Lately we moved away from the `ErrorHandler` also doing the termination and split it into a reporting class and a separate `panic` functionality. The reporting would be simple and without the three levels we currently have. Just something like
```cpp
IOX_REPORT(FOO_ERROR);
// or maybe more generic
IOX_STATUS(FOO)
PANIC;
```
Maybe even something like this
```cpp
IOX_STATUS(FOO, FATAL) << "With a log message"; // the second parameter would always be the log level
PANIC;
```

The only problem would be where to put `PANIC`? Into `primitives`? Create an `assertions` module?

Reply:
@elBoberido As I was the one who established the term "ErrorReporting" a couple of months ago, I like the idea :smile: `PANIC` going into `primitives` makes sense to me.